### ISSUE TITLE
MD025 rule incorrectly flags code comments and pre-compiler directives in fenced code blocks as H1 headers

### ISSUE THREAD

**Issue by huckncatch** [DavidAnson/markdownlint-cli2#726]
## Description

The MD025 rule (single-title/single-h1) is incorrectly identifying lines starting with `# ` inside fenced code blocks as H1 headers, causing false positive violations. This affects various programming languages that use `#` for comments, pre-compiler directives, and other syntax.

## Expected Behavior

Content inside fenced code blocks should not be parsed as markdown headers, regardless of the `#` character usage. Only actual markdown headers (lines starting with `# ` outside of code blocks) should trigger the MD025 rule.

## Actual Behavior

The extension treats every line starting with `# ` as an H1 header, even when they are legitimate code syntax inside properly fenced code blocks with language identifiers. The lines inside code blocks are not flagged.

## Steps to Reproduce

1. Create a markdown file with this content:

````markdown
# Main Title

Some text here.

```bash
# This is a bash comment
echo "Hello World"
```

```swift
#if DEBUG
import Inject
#endif

struct ContentView: View {
    var body: some View {
        Text("Hello, World!")
        #if DEBUG
            .injectReload()
        #endif
    }
}
```

```c
#include <stdio.h>
#define MAX_SIZE 100

int main() {
    # This is also a comment in some contexts
    return 0;
}
```

More text here.
````

2. Lint the file with markdownlint-cli2
3. Observe the MD025 violation is listed.

## Affected Languages

This issue impacts documentation for many programming languages:
- **Bash/Shell**: Comments (`# comment`)
- **Swift**: Pre-compiler directives (`#if DEBUG`, `#endif`, `#import`)
- **C/C++**: Pre-processor directives (`#include`, `#define`, `#ifdef`)
- **Python**: Comments (`# comment`)
- **Ruby**: Comments (`# comment`)
- **Perl**: Comments (`# comment`)
- **And many others...**

## Environment

- **OS:** macOS
- **markdownlint-cli2 version:** v0.19.1 (markdownlint v0.39.0)

## Additional Context

This issue particularly affects technical documentation, programming tutorials, and setup guides where code blocks frequently contain comments and pre-compiler directives. The false positives make the MD025 rule difficult to use effectively for its intended purpose of ensuring single main titles per document.

The parser should respect fenced code block boundaries and not apply markdown header rules to their contents.

**Comment by DavidAnson** [comment]
The documentation explains this behavior. The title in your front matter counts as the first top level heading, so the one at the beginning of the document is reported as a second. The documentation also explains how to disable this.

https://github.com/DavidAnson/markdownlint/blob/main/doc/md025.md
